external/dhcpd/dhcpd_lwip.c

```diff
@@ -109,66 +174,86 @@
 int dhcp_server_status(char *intf)
 {
-	struct netif *cur_netif;
-	cur_netif = netif_find(intf);
-	if (cur_netif == NULL) {
-		ndbg("[DHCPS] no network interface for dhcpd start\n");
-		return 0;
+	int ret = ERROR;
+	struct req_lwip_data req;
+
+	int sockfd = socket(AF_INET, SOCK_DGRAM, 0);
+	if (sockfd < 0) {
+		ndbg("socket() failed with errno: %d\n", errno);
+		return ret;
 	}
 
-	if (cur_netif->dhcps_pcb == NULL) {
-		ndbg("[DHCPS] DHCP server closed\n");
-		return 0;
-	} else {
-		ndbg("[DHCPS] DHCP server opened\n");
-		return 1;
+	memset(&req, 0, sizeof(req));
+	req.type = DHCPDSTATUS;
+	req.host_name = intf;
+
+	ret = ioctl(sockfd, SIOCLWIP, (unsigned long)&req);
+	if (ret == ERROR) {
+		ndbg("ioctl() failed with errno: %d\n", errno);
+		close(sockfd);
+		return ret;
 	}
+
+	ret = req.req_res;
+	close(sockfd);
+	return ret;
 }
 
 /****************************************************************************
  * Name: dhcps_server_start
  ****************************************************************************/
+
+
 int dhcp_server_start(char *intf, dhcp_sta_joined dhcp_join_cb)
 {
-	struct netif *cur_netif;
-	cur_netif = netif_find(intf);
-	if (cur_netif == NULL) {
-		ndbg("[DHCPS] no network interface for dhcpd start\n");
-		return -1;
-	}
-	if (dhcp_join_cb) {
-		if (dhcps_register_cb(dhcp_join_cb) != ERR_OK) {
-			ndbg("[DHCPS] link callback fail\n");
-			return -1;
-		}
-	}
+	dhcp_join_data_s *data = (dhcp_join_data_s *)malloc(sizeof(dhcp_join_data_s));
+	data->intf = (char *)malloc(strlen(intf)+1);
+	memcpy(data->intf, intf, strlen(intf) + 1);
+	data->fn = dhcp_join_cb;
 
-	if (netifapi_dhcps_start(cur_netif) == ERR_OK) {
-		ndbg("[DHCPS] started successfully (LWIP)\n");
-		return OK;
+	int ret = ERROR;
+	g_dhcpd_term = 0;
+	ret = pthread_create(&g_dhcpd_tid, NULL, _dhcpd_join_handler, (void *)data);
+	pthread_setname_np(g_dhcpd_tid, "dhcpd cb handler");
+	if (ret < 0) {
+		g_dhcpd_term = 1;
+		ndbg("[dhcpd] create iotapi handler fail(%d) errno %d\n", ret, errno);
+		return ERROR;
 	}
-
-	return -1;
+	return ret;
 }
 
 /****************************************************************************
  * Name: dhcps_server_stop
  ****************************************************************************/
 int dhcp_server_stop(char *intf)
 {
-	struct netif *cur_netif;
-	cur_netif = netif_find(intf);
-	if (cur_netif == NULL) {
-		ndbg("[DHCPS] no network interface for dhcpd start\n");
-		return -1;
+	int ret = ERROR;
+	struct req_lwip_data req;
+	g_dhcpd_term = 1;
+	sleep(1);
+
+	pthread_join(g_dhcpd_tid, NULL);
+
+	int sockfd = socket(AF_INET, SOCK_DGRAM, 0);
+	if (sockfd < 0) {
+		ndbg("socket() failed with errno: %d\n", errno);
+		return ret;
 	}
-	if (cur_netif->dhcps_pcb == NULL) {
-		ndbg("[DHCPS] stop dhcpd fail: no pcb\n");
-		return -1;
+
+	memset(&req, 0, sizeof(req));
+	req.type = DHCPDSTOP;
+	req.host_name = intf;
+
+	ret = ioctl(sockfd, SIOCLWIP, (unsigned long)&req);
```

Comment:
I gave to you review already but how application layer knows ip layer's feature?? Of course file name including 'lwip'
But it is ioctl's command definition, this should be handled above of ip layer, I think
You can define SIOC_DHCPSTART / SIOC_DHCPSTOP and so on in ioctl.h
then socket should check current ip type and call functions properly.

Reply:
Thanks for the comment. This is an application, but it is kind of system utils. Based on the configuration, this application can use either 1) external dhcp or 2) LwIP (kernel) 's dhcp. For considering DHCP of LwIP, this structure is inevitable. As you know, LwIP has very exceptional architecture, so there is no other options, without changing the whole structure.